src/solvers/strings/string_refinement.cpp

```diff
@@ -1200,12 +1206,24 @@ static void substitute_array_access_in_place(
 {
   if(const auto index_expr = expr_try_dynamic_cast<index_exprt>(expr))
   {
-    expr =
+    optionalt<exprt> result =
       substitute_array_access(*index_expr, symbol_generator, left_propagate);
+
+    // Only perform a write when we have something changed.
+    if(result)
+    {
+      expr.write();
+      expr.swap(*result);
+    }
   }
 
-  for(auto &op : expr.operands())
-    substitute_array_access_in_place(op, symbol_generator, left_propagate);
+  // Trick it into giving us a const iterable (thus not copying anything) as we
```

Comment:
Rather than using const-hacking, use the depth-iterator, which has a `mutate()` operation for exactly this purpose. It would look something like
```
for(auto it = expr.depth_begin(), itend = expr.depth_end(); it != itend; ++it)
{
  if(const auto index_expr = expr_try_dynamic_cast<index_exprt>(*it))
  {
    optionalt<exprt> result =
￼       substitute_array_access(*index_expr, symbol_generator, left_propagate);		￼       ￼     
      if(result)
        it.mutate() = *result;
  }
}
```
That performs CoW breaks from the root of the expression up to the modified node on demand.

Reply:
:+1: but I'm unsure about `itend` being defined at the beginning of a loop which can mutate `expr`